### PR TITLE
Remove syncthing process based on name

### LINKED ISSUE #1204: "okteto up" fails with another syncthing instance running error
**Describe the bug**

`okteto up` fails and the local syncthing logs "Error opening database: resource temporarily unavailable (is another instance of Syncthing running?)".

**Expected behavior**
`okteto up` should be able to recover from this error.

### PR DESCRIPTION
**Pull request by pchico83**
Fixes #1204

### CODE REVIEW

File: pkg/syncthing/process.go [hunk 1]
```diff
@@ -16,22 +14,47 @@
...
+	"github.com/shirou/gopsutil/process"
 )
 
-func terminate(pid int, wait bool) error {
-	proc := os.Process{Pid: pid}
-	if err := proc.Signal(os.Interrupt); err != nil {
-		if strings.Contains(err.Error(), "process already finished") {
-			return nil
-		}
+func getParentExe(p *process.Process) string {
```
Comment (rberrelleza) [inline pkg/syncthing/process.go:23]: if this just for windows? If this code depends on windows components, you need to put it in the  process_windows.go folder, so it doesn't get compiled on Linux or macos (it's based on build tag at the top of the file)

File: pkg/syncthing/syncthing.go [hunk 2]
```diff
@@ -702,32 +668,34 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
+	for _, p := range pList {
+		exe, err := p.Exe()
+		if err != nil {
+			log.Infof("error getting procress exec: %s", err.Error())
+			continue
+		}
+		if exe != getInstallPath() {
+			continue
+		}
+		if getParentExe(p) == getInstallPath() {
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:685]: what is parentexe on this context?

File: pkg/syncthing/syncthing.go [hunk 3]
```diff
@@ -702,32 +668,34 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
-	if !force {
-		if pid != s.pid {
-			log.Infof("syncthing pid-%d wasn't created by this command, skipping", pid)
-			return nil
-		}
-	}
-
-	if err := s.cleanupDaemon(pid, force); err != nil {
+func (s *Syncthing) Stop(wait bool) error {
+	pList, err := process.Processes()
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:672]: in my experience, listing processes and waiting on them is a bit flaky and time consuming, do you think we should only do this if the pidPath is not valid?

Comment (rberrelleza) [reply pkg/syncthing/syncthing.go:672]: can you also validate if process.Processes() works when you're not an admin on the machine? (it's something common on windows environments)

File: pkg/syncthing/process.go (hunk 1 above)
Comment (rberrelleza) [reply pkg/syncthing/process.go:23]: Is not clear what 'Exe' means in this context, maybe rename to something more explicit? (I believe this is returning the path to the parent of `p` ?

File: pkg/syncthing/process.go [hunk 4]
```diff
@@ -16,22 +14,47 @@
...
+}
 
+func terminate(p *process.Process, wait bool) error {
+	if err := p.Terminate(); err != nil {
 		return err
 	}
 
 	if wait {
-		defer proc.Wait() // nolint: errcheck
+		isRunning, err := p.IsRunning()
```
Comment (rberrelleza) [inline pkg/syncthing/process.go:47]: double check, but I believe p.IsRunning doesn't work as expected in windows if the process doesn't exist.

File: pkg/syncthing/syncthing.go [hunk 5]
```diff
@@ -702,32 +668,36 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
+	if err != nil {
 		return err
 	}
-
-	if err := os.Remove(pidPath); err != nil {
-		if os.IsNotExist(err) {
-			return nil
+	for _, p := range pList {
+		exe, err := p.Exe()
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:677]:
```suggestion
		executablePath, err := p.Exe()
```

File: pkg/syncthing/syncthing.go [hunk 6]
```diff
@@ -702,32 +668,36 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
+		if err != nil {
+			log.Infof("error getting procress exec: %s", err.Error())
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:679]: add the pid or process name so it helps troubleshoot?
```suggestion
			log.Infof("error getting exec path for process %s: %s", p.Pid, err.Error())
```

File: pkg/syncthing/syncthing.go [hunk 7]
```diff
@@ -702,32 +668,36 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
+			continue
 		}
-
-		log.Infof("failed to delete pidfile %s: %s", pidPath, err)
+		if exe != getInstallPath() {
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:682]: can you add comments to the intention of the check? It's clear from just looking at the code (I'm guessing you're validating if this is syncthing by checking the path?)

File: pkg/syncthing/syncthing.go (hunk 2 above)
Comment (rberrelleza) [reply pkg/syncthing/syncthing.go:685]: why are we skipping if it's the parent?

File: pkg/syncthing/syncthing.go [hunk 8]
```diff
@@ -702,32 +668,36 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
+			continue
+		}
+		if getParentExe(p) == getInstallPath() {
+			continue
+		}
+		argLine, err := p.Cmdline()
+		if err != nil {
+			return err
+		}
+		if !strings.Contains(argLine, fmt.Sprintf("-home %s", s.Home)) {
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:692]: can you add a comment with the expected argLine?

File: pkg/syncthing/syncthing.go [hunk 9]
```diff
@@ -702,32 +668,36 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
+			continue
+		}
+		log.Infof("terminating syncthing with wait %t", wait)
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:695]: can you add which synching process you're terminating?  pid, path, args? that can help us troubleshoot if this doesn't work as expected.

File: pkg/syncthing/syncthing.go [hunk 10]
```diff
@@ -702,32 +668,36 @@ func (s *Syncthing) Restart(ctx context.Context) error {
...
-
-	if !force {
-		if pid != s.pid {
-			log.Infof("syncthing pid-%d wasn't created by this command, skipping", pid)
-			return nil
-		}
-	}
-
-	if err := s.cleanupDaemon(pid, force); err != nil {
+func (s *Syncthing) Stop(wait bool) error {
```
Comment (rberrelleza) [inline pkg/syncthing/syncthing.go:671]: we should add tests for this, especially to validate that we're not killing syncthing's owned by a different up, or other processes running in the machine.

**Review by rberrelleza — COMMENTED** [review]
looks good, we should test in windows.

File: pkg/syncthing/syncthing.go (hunk 3 above)
Comment (pchico83) [reply pkg/syncthing/syncthing.go:672]: It is slower. I think I am going to do it just at the beginning of the `up` sequence and in parallel with the rest of the `up` sequence.
For windows, yes, it also works for non-admin users

File: pkg/syncthing/syncthing.go (hunk 2 above)
Comment (pchico83) [reply pkg/syncthing/syncthing.go:685]: This is to skip the child of the main syncthing process, which should be terminated by the parent

**Pull request merged**